### PR TITLE
Narrow utf16 to ascii

### PR DESCRIPTION
**Pull request by pgovind**
This is ready for review now. 

NOTE: There is a bug here somewhere, but I can't seem to find it, so I'd appreciate another pair of eyes in this code. When I try to test this on WSL2, I get an error saying "Can't find file xunit.console.deps.json" when that file does in fact exist. The real error is that the xunit infra itself calls into the method being modified here and **something** goes wrong (maybe marshalling the path from `string` to native also uses this code? I'm not sure). AFAICT the changes are right.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs [hunk 1]
```diff
@@ -215,6 +215,17 @@ private static unsafe nuint GetIndexOfFirstNonAsciiByte_Default(byte* pBuffer, n
             goto Finish;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool ContainsNonAsciiValue(Vector128<ushort> value)
+        {
+            if (!AdvSimd.Arm64.IsSupported)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs:221]: Maybe add a little endian assertion check here, for self-documentation purposes?

File: src/libraries/System.Utf8String.Experimental/src/System/Runtime/Intrinsics/Intrinsics.Shims.cs [hunk 2]
```diff
@@ -9,6 +9,8 @@ internal static class Vector64
         public static Vector64<uint> CreateScalar(uint value) => throw new PlatformNotSupportedException();
         public static Vector64<byte> AsByte<T>(this Vector64<T> vector) where T : struct => throw new PlatformNotSupportedException();
         public static Vector64<uint> AsUInt32<T>(this Vector64<T> vector) where T : struct => throw new PlatformNotSupportedException();
+        public static Vector64<T> GetLower<T>(this Vector128<T> vector) where T : struct => throw new PlatformNotSupportedException();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Utf8String.Experimental/src/System/Runtime/Intrinsics/Intrinsics.Shims.cs:12]: Is this for netstandard2.0? Do we not have a shimmed vector file in corelib?

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs [hunk 3]
```diff
@@ -1322,16 +1344,43 @@ private static unsafe nuint NarrowUtf16ToAscii_Sse2(char* pUtf16Buffer, byte* pA
...
                 {
-                    return 0;
+                    if ((Sse2.MoveMask(Sse2.AddSaturate(utf16VectorFirst.AsUInt16(), asciiMaskForAddSaturate).AsByte()) & NonAsciiDataSeenMask) != 0)
+                    {
+                        return 0;
+                    }
+                }
+                else if (AdvSimd.Arm64.IsSupported)
+                {
+                    if (ContainsNonAsciiValue(AdvSimd.AddSaturate(utf16VectorFirst.AsUInt16(), asciiMaskForAddSaturate)))
```
Comment (echesakovMSFT) [inline src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs:1356]: If I understand the original Sse2 algorithm right - this is testing if any of the characters in `utf16VectorFirst` is outside of `[0, 0x7f]` value range, right?

I don't think the algorithm on Arm64 should copy exactly, what Sse2 does. 

I believe the following should be enough
```c#
AdvSimd.MaxPairwise(utf16VectorFirst.AsUInt16(), utf16VectorFirst.AsUInt16()).AsUInt64().ToScalar() & 0xFF80FF80FF80FF80UL != 0
```

File: src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs [hunk 4]
```diff
@@ -1298,17 +1307,30 @@ private static unsafe nuint NarrowUtf16ToAscii_Sse2(char* pUtf16Buffer, byte* pA
...
-            Vector128<short> utf16VectorFirst = Sse2.LoadVector128((short*)pUtf16Buffer); // unaligned load
+            Vector128<short> utf16VectorFirst;
+
+            if (Sse2.IsSupported)
+            {
+                utf16VectorFirst = Sse2.LoadVector128((short*)pUtf16Buffer); // unaligned load
+            }
+            else if (AdvSimd.Arm64.IsSupported)
+            {
+                utf16VectorFirst = AdvSimd.LoadVector128((short*)pUtf16Buffer);
```
Comment (kunalspathak) [inline src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs:1328]: minor: add a comment ` // unaligned load`

File: src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs (hunk 3 above)
Comment (kunalspathak) [reply src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs:1356]: I agree.

**Comment by jeffhandley** [comment]
@pgovind I'm going to close this PR since this work is on hold. When we get back around to this work, we can create a fresh PR.

**Pull request closed without merging**